Add extension methods to package:collection.

lrhn opened the pull request:
Add a number of useful extension methods to `List` and `Iterable`, and to a few other types.

A good part of these extensions are backed by the algorithms of `algorithms.dart`, which is expanded to support them:

* Added `quickSort`.
* Add extra optional `Random` argument to `shuffle`.
* Generalize some of the functions in `algorithms.dart` to work on specific properties of the objects (`binarySearchBy`,
  `lowerBoundBy`, `insertionSortBy`, `quickSortBy`, `mergeSortBy`).

lrhn commented:
This is a first crack at adding extension methods to package:collection.

They are not *tested*. Some are probably not useful (I expect `Iterable<T extends num>.sum()` to be mostly speculative).

natebosch commented on analysis_options.yaml:
```diff
@@ -1,42 +1,6 @@
-    - type_init_formals
-    - unnecessary_const
-    - unnecessary_new
+    annotate_overrides: ignore
```
We should be following this lint.

lrhn replied:
Done for now.

natebosch commented on analysis_options.yaml:
```diff
@@ -1,42 +1,6 @@
-    - unnecessary_new
+    annotate_overrides: ignore
+    prefer_single_quotes: ignore
+    use_function_type_syntax_for_parameters: ignore
```
We should be following this lint.

Some of the removed ones that aren't in `pedantic` yet are also useful. We rarely violate them on the team, but I don't see any reason for something like `camel_case_types` or `library_prefixes` to need to wait for code review if we do get a contribution from outside the team.

lrhn replied:
As you may  be aware, I disagree with following that lint. It's not part of the Dart style guide, and I actually prefer the inline notation for function arguments, both for reading and writing.
These are the three exceptions that I apply because I consistently write code that does not follow them.

I'm using pedantic+modifications because it's convenient (there are good things in there too, just not all of it), not because I think it's what we actually want as the Dart recommended lint set outside of Google3.
(Sure, I can add camel_case_types, but that has never been an issue.)

jakemac53 replied:
How is it possibly a good thing to have every random person who touches a codebase changing the lints to fit their personal preference?

natebosch replied:
> It's not part of the Dart style guide, and I actually prefer the inline notation for function arguments, both for reading and writing.

https://dart.dev/guides/language/effective-dart/design#consider-using-function-type-syntax-for-parameters

It is a "Consider" it think for largely historical reasons. In future revisions it's likely to become stronger. The fact that there _are_ 2 syntaxes for this will hopefully be fixed in Dart 3 or some 2.x language version - ideally we have 1 way to do things instead of 2 competing ways that require constant negotiation or inconsistent codebases.

> These are the three exceptions that I apply because I consistently write code that does not follow them.

This goes against the flow of the rest of Google Dart authors, and against the general philosophy of Readability in Google.

> Sure, I can add camel_case_types, but that has never been an issue.

It has never been a _big_ issue. It has come up in code reviews in the past and the intention of putting it in a lint is to save time, and save from re-negotiating it with new contributors. The reason `camel_case_types` isn't in `pedantic` is because that would require consistent enforcement, and in internal code I think we have some exceptions. We won't have exceptions in _this_ package which is why the lint is enabled.


I know that we have different approaches to linting in different places. I don't think having us each try to assert our own style into shared packages is the right thing to do, neither is introducing config changes into other PRs.

natebosch commented on lib/src/empty_unmodifiable_set.dart:
```diff
@@ -6,15 +6,10 @@ import 'dart:collection';
-// can't use mixins.
 /// An unmodifiable, empty set that can have a const constructor.
 class EmptyUnmodifiableSet<E> extends IterableBase<E>
+    with UnmodifiableSetMixin<E>
```
This seems like a good change, but I don't understand how it's related. Can this go out in a separate commit?

lrhn replied:
It's an unrelated drive-by fix. I can do that by itself too.

natebosch replied:
> I can do that by itself too.

I think that would be better. The travis wait time for individual PRs is the only downside - if you need to send a few things together to work around that I think it's OK, otherwise for isolated refactorings I think we should follow go/small-cls

natebosch commented on lib/src/algorithms.dart:
```diff
@@ -16,14 +19,24 @@ import 'utils.dart';
+/// property of the elements, the result is unpredictable.
+///
+/// Returns -1 if [value] is not in the list by default.
+int binarySearchBy<E, K>(List<E> sortedList, K keyOf(E element), int Function(K, K) compare, E value) {
```
I don't think it's a good idea to add new top level methods to this package. Anything new should be added as an extension only. If we want to have these as top level methods internally for some reason, we should be hiding them from our exports.

lrhn replied:
I completely disagree with not adding top-level functions.

Top-level functions are still a perfectly good way to introduce general reusable functionality.

Extension methods are less generally useful than top-level functions, but more convenient and discoverable.
That's why I have made the distinction of adding convenient and discoverable APIs using extension methods, and more rough and verbose reusable APIs as top-level functions.

natebosch replied:
> Not every function deserves looking like an instance-method.

When I last looked it seemed like every function in _this_ packages deserves looking like an instance-method. If we wont' be adding `list.binarySearchBy` extension at all, that does change the discussion a bit, but I think that would be a good extension method.

> It also avoids the "all functions after 2.6 are extension methods, all functions before are not" issue.

I don't think that is an issue, I haven't seen anyone make that mistake.

The main thing I want to push on is to have 1 (recommended) way to do things. I want the users of this package to avoid negotiating during _their_ code review about whether to use the extension method version or the top-level method version.

> The general function doesn't really _fit in_ as an instance/extension member, but it's there in case you really need the full power.

I don't understand this. Is this ability to be passed as a function (with a _super_ specific function signature) rather than be torn off a specific list the "full power"?

lrhn replied:
There are multiple layers of API here. Both a general low-level API that people can use to build other libraries from, and which isn't necessarily optimized for ease-of-use, and a high-level convenience API that accesses the low-level API in various convenient manners. Actually, most of the former until now. It's a package of building blocks more than a package of solutions.

We could stop exposing the low-level API directly. That would mean not exporting the new quickSort or xBy methods from `algorithms` and only expose them as extension methods. It'll probably work. There will likely be someone asking us to expose it anyway, but if the full power is still accessible as an extension method, it's probably not an issue.

As for `binarySearch`, the reasons I originally didn't make it as an extension is that it doesn't actually work on all lists. It only works on *sorted* lists, and therefore I didn't want it to be an instance method on all lists. (For the same reason I think `sort` is badly designed as well, but that can be saved by extension methods, binary search cannot).
It's ... probably overly cautious of me, and I'll add the binary search as extension methods as well.

natebosch replied:
> There will likely be someone asking us to expose it anyway, but if the full power is still accessible as an extension method, it's probably not an issue.

Waiting to expose something until someone asks is my preferred approach. As a team we've spent a non-trivial amount of time trying to remove stuff that has little or no usage.

lrhn replied:
We're now not exposing the new top-level members.

natebosch reviewed (changes requested):
Did something get lost in the upload? There are no extension methods added here. This does look like it might be some work towards adding extensions, but it doesn't match the PR description as is.

lrhn commented:
Whoops. Forgot to git-add the new files.

natebosch commented on lib/src/iterable_extensions.dart:
```diff
@@ -0,0 +1,519 @@
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// Extension methods on iterables.
```
This isn't a useful comment.

natebosch commented on lib/src/iterable_extensions.dart:
```diff
@@ -0,0 +1,519 @@
+
+import "algorithms.dart";
+
+/// Extensions that apply to all iterables.
```
If the only thing we can say is to repeat the signature, we should omit the comment. Same thing for all the docs on the extensions in this file.

natebosch commented on lib/src/iterable_extensions.dart:
```diff
@@ -0,0 +1,519 @@
+  /// Creates a sorted list of the elements of the iterable.
+  ///
+  /// The elements are ordered by the [compare] [Comparator].
+  List<T> sorted(int compare(T a, T b)) => [...this]..sort(compare);
```
Should we match the signature on `List.sort` and make the argument optional?

lrhn replied:
The signature of `List.sort` is actually bad. It allows you to omit the `compare`, but that only really works when the elements are `Comparable`. With extension methods we can do *better*.
There is a `sorted([int compare(E a, E b)])` on `Iterable<T extends Comparable<T>>`.

natebosch commented on lib/src/iterable_extensions.dart:
```diff
@@ -0,0 +1,519 @@
+  }
+
+  /// Whether the elements are sorted by the [compare] ordering of [keyOf].
+  bool isSortedBy<K>(K keyOf(T element), int compare(K a, K b)) {
```
This is inconsistent with `sortedBy` and `sortedCompareBy`. Should we rename this `sortedCompareBy` and add a `sortedBy<K extends Comparable<K>>`?

lrhn replied:
Seems reasonable.
And it should take start/end offsets too.

natebosch commented on lib/src/iterable_extensions.dart:
```diff
@@ -0,0 +1,519 @@
+    return true;
+  }
+
+  /// The elements of this iterable with duplicates removed.
```
Does this offer enough value over `.toSet()` to make it worth adding?

lrhn replied:
Probably not, no.

natebosch commented on lib/src/iterable_extensions.dart:
```diff
@@ -0,0 +1,519 @@
+
+  /// The elements of this iterable with duplicates removed.
+  Iterable<T> unique() sync* {
+    var seen = <T>{};
```
Would `HashSet<T>();` be cheaper? We don't care about maintaining insertion order.

natebosch commented on lib/src/iterable_extensions.dart:
```diff
@@ -0,0 +1,519 @@
+    var result = <K, Set<T>>{};
+    for (var element in this) {
+      var key = keyOf(element);
+      result[key] = (result[key] ?? {})..add(element);
```
I'm not sure if it actually makes a difference (or if we should be trying to micro-optimize without benchmarks) but I _think_ this is faster on dart2js with `??=`

```suggestion
      (result[key] ??= {}).add(element);
```

natebosch commented on lib/src/iterable_extensions.dart:
````diff
@@ -0,0 +1,519 @@
+  /// var parts = [1, 2, 3, 4, 5, 6, 7, 8, 9].split(isPrime);
+  /// print(parts); // ([1], [2], [3, 4], [5, 6], [7, 8, 9])
+  /// ```
+  Iterable<List<T>> splitBefore(bool test(T element)) =>
````
Is this speculative, or has someone asked for this?

natebosch commented on lib/src/iterable_extensions.dart:
```diff
@@ -0,0 +1,519 @@
+    return result;
+  }
+
+  /// The average of the elements.
```
We need to call out that this is the mean, not the median. Or we could name it `mean` or `arithmeticMean` and avoid the ambiguous `average` term altogether.

natebosch commented on lib/src/iterable_extensions.dart:
```diff
@@ -0,0 +1,519 @@
+  /// The sum of the elements.
+  ///
+  /// The sum is zero if the iterable is empty.
+  T sum() {
```
Should these be getters?

natebosch commented on lib/src/iterable_extensions.dart:
```diff
@@ -0,0 +1,519 @@
+  /// natural [Comparable] ordering of the elements.
+  List<T> sorted([int compare(T a, T b)]) => [...this]..sort(compare);
+
+    /// Whether the elements are sorted by the [compare] ordering.
```
The doc should describe the case where `compare` is omitted.

```dart
  /// Whether the element are sorted.
  ///
  /// The order is checked against [compare] if it is passed,
  /// otherwise it is checked against the element's natural order.
```

natebosch commented on lib/src/list_extensions.dart:
```diff
@@ -0,0 +1,281 @@
+  }
+
+  /// Maps each element and its index to a new value.
+  Iterable<R> mapIndexed<R>(R convert(int index, E element)) sync* {
```
How much value are we buying by specializing this against `Iterable.mapIndexed`? Do we expect `this[index]` to be meaningfully different (faster?) than `for(var element in this)`?

lrhn replied:
Yes, but maybe not by a lot.

natebosch commented on lib/src/list_extensions.dart:
```diff
@@ -0,0 +1,281 @@
+  }
+
+  /// Expands each element and index to a number of elements in a new iterable.
+  Iterable<R> expendIndexed<R>(Iterable<R> expend(int index, E element)) sync* {
```
Typo

```suggestion
  Iterable<R> expandIndexed<R>(Iterable<R> expend(int index, E element)) sync* {
```

natebosch commented on lib/src/list_extensions.dart:
```diff
@@ -0,0 +1,281 @@
+  }
+
+
+  /// Sort a range of elements by [compare].
```
I don't think we should add both this _and_ the ability to sort a `ListSlice`

natebosch commented on lib/src/list_extensions.dart:
```diff
@@ -0,0 +1,281 @@
+  /// Whether [other] has the same elements as this list.
+  ///
+  /// Returns true iff [other] has the same [length]
+  /// as this list, and the elemets of this list and [other]
```
Typo
```suggestion
  /// as this list, and the elements of this list and [other]
```

natebosch commented on lib/src/list_extensions.dart:
```diff
@@ -0,0 +1,281 @@
+  ///
+  /// Returns true iff [other] has the same [length]
+  /// as this list, and the elemets of this list and [other]
+  /// at the same indices are equal (according to `==`).
```
Comment needs to be rephrased to account for the `equality` argument.

mit-mit commented:
@lrhn looks stale?

lrhn commented:
PTAL

natebosch commented:
I had been waiting to review after the merge conflicts were resolved... Do you want me to do a first pass and then re-review after solving the conflicts?

lrhn commented:
Conflicts should be gone now.

natebosch commented on lib/src/iterable_extensions.dart:
```diff
@@ -0,0 +1,725 @@
+  /// Creates a sorted list of the elements of the iterable.
+  ///
+  /// The elements are ordered by the [compare] [Comparator].
+  List<T> sorted(Comparator<T> compare) => [...this]..sort(compare);
```
Should we make this an optional argument to match `List.sort`?

natebosch replied:
I forgot I had already asked this and we resolved the conversation. Sorry about that!

https://github.com/dart-lang/collection/pull/135#discussion_r425658615

natebosch commented on lib/src/iterable_extensions.dart:
```diff
@@ -0,0 +1,725 @@
+  /// Whether the elements are sorted by the [compare] ordering.
+  bool isSorted(Comparator<T> compare) {
+    var iterator = this.iterator;
+    if (!iterator.moveNext()) return true;
```
[nit] I think it would be worth calling out the behavior for an empty list in the doc comment.

natebosch commented on lib/src/iterable_extensions.dart:
```diff
@@ -0,0 +1,725 @@
+  /// The single element satisfying [test].
+  ///
+  /// Returns `null` if there are either no elements
+  /// or more than one element satisfying [test].
```
In the SDK I think the `singleWhere` methods throw if there is more than one match, and use `orElse` only for the no-match case. This is a different behavior that may be a better design, but also could be confusing.

If we do want this behavior, can we call out the difference specifically?

natebosch commented on lib/src/iterable_extensions.dart:
```diff
@@ -0,0 +1,725 @@
+  /// The arithmetic mean is the sum of the elements
+  /// divided by the number of elements.
+  ///
+  /// The list must not be empty.
```
[nit] Replace all "list" in the comment with "iterable"

natebosch commented on lib/src/iterable_extensions.dart:
```diff
@@ -0,0 +1,725 @@
+  /// divided by the number of elements.
+  ///
+  /// The list must not be empty.
+  double get average {
```
Should we go with an algorithm that won't overflow?

```dart
extension on Iterable<num> {
  double get average {
    var average = 0.0;
    var count = 1;
    for (var element in this) {
      average += (element - average) / count;
      count++;
    }
    return average;
  }
}
```

natebosch commented on lib/src/iterable_extensions.dart:
```diff
@@ -0,0 +1,725 @@
+/// Extensions on comparator functions.
+extension ComparatorExtension<T> on Comparator<T> {
+  /// The inverse ordering of this comparator.
+  int Function(T, T) get inverse => (T a, T b) => this(b, a);
```
[nit] Why use the function type here and the name `Comparator` below?

natebosch reviewed (approved):
Overall I'm _slightly_ worried about adding so many of these speculatively since there is a high cost to remove them, but given that they are extension methods I think the overall cost to keeping them is low.

lrhn committed:
Add extension methods to package:collection.

Add a number of useful extension methods to List and Iterable, and to a few other types.
Also add a quick-sort implementation that these can extension methods can defer to,
one which allows sorting only a range of the list.

lrhn commented on CHANGELOG.md:
```diff
@@ -3,6 +3,20 @@
+
+* Add extra optional `Random` argument to `shuffle`.
+
+* Generalize some of the functions in `algorithms.dart` to
```
Done.  Also removing quickSort reference above, since it's also not exported.

lrhn committed:
Address comments

lrhn reviewed (commented):
Addressed comments.
